metis-harvesting/src/main/java/eu/europeana/metis/harvesting/oaipmh/OaiRecordParser.java

```diff
@@ -79,58 +81,48 @@ private static void initializeExpressions() throws HarvesterException {
    * @param oaiRecord The record to parse as a string value.
    * @throws HarvesterException In case there was a problem with setting up the parser.
    */
-  OaiRecordParser(String oaiRecord) throws HarvesterException {
+  OaiRecordParser(byte[] oaiRecord) throws HarvesterException {
     initializeExpressions();
     this.oaiRecord = oaiRecord;
   }
 
   /**
-   * Obtain the embedded RDF record
+   * Obtain the embedded (RDF) record with it's OAI header.
    *
-   * @return The embedded RDF record as a stream.
+   * @return The record along with the OAI header.
    * @throws HarvesterException in case there is a problem with the expression.
    */
-  InputStream getRdfRecord() throws HarvesterException {
+  OaiRecord getOaiRecord() throws HarvesterException {
     try {
-      final InputSource inputSource = getInputSource();
-      // Note that this is created safely, so this is a false positive by SonarQube.
-      @SuppressWarnings("findsecbugs:XXE_XPATH")
-      final ArrayList<NodeInfo> result = (ArrayList<NodeInfo>) metadataExpression
-              .evaluate(inputSource, XPathConstants.NODESET);
-      return convertToStream(result);
-    } catch (XPathExpressionException | TransformerException | RuntimeException e) {
-      throw new HarvesterException("Cannot xpath XML!", e);
-    }
-  }
-
-  /**
-   * Find whether the record is marked as deleted.
-   *
-   * @return whether the method is marked as deleted.
-   * @throws HarvesterException in case there is a problem with the expression.
-   */
-  boolean recordIsDeleted() throws HarvesterException {
-    try {
-      // Note that this is created safely, so this is a false positive by SonarQube.
-      @SuppressWarnings("findsecbugs:XXE_XPATH")
-      final String result = isDeletedExpression.evaluate(getInputSource());
-      return "deleted".equalsIgnoreCase(result);
-    } catch (XPathExpressionException e) {
+      final byte[] headerBytes = getAsBytes(headerExpression);
```

Comment:
If the constructor is updated to use the same static XPathExpression for all records I think those XPathExpression should be used with synchronization to avoid weird issues we had before with those classes(I think it depends on the  exact implementation of the library that is used, but we might want to stay on the safe side).

Reply:
OK. Done.